pkg/infrastructure/clusterapi/clusterapi.go

```diff
@@ -0,0 +1,317 @@
+package clusterapi
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/sirupsen/logrus"
+	"gopkg.in/yaml.v2"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	utilkubeconfig "sigs.k8s.io/cluster-api/util/kubeconfig"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/manager/signals"
+
+	"github.com/openshift/installer/pkg/asset"
+	"github.com/openshift/installer/pkg/asset/cluster/metadata"
+	"github.com/openshift/installer/pkg/asset/ignition/bootstrap"
+	"github.com/openshift/installer/pkg/asset/ignition/machine"
+	"github.com/openshift/installer/pkg/asset/installconfig"
+	"github.com/openshift/installer/pkg/asset/kubeconfig"
+	"github.com/openshift/installer/pkg/asset/manifests/capiutils"
+	capimanifests "github.com/openshift/installer/pkg/asset/manifests/clusterapi"
+	"github.com/openshift/installer/pkg/clusterapi"
+	"github.com/openshift/installer/pkg/infrastructure"
+	"github.com/openshift/installer/pkg/types"
+)
+
+// ClusterAPIProvider is the base implementation for provisioning cluster
+// infrastructure using CAPI. Platforms should embed this struct and
+// implement:
+// .
+type InfraProvider struct {
+	infrastructure.Provider
+
+	capiProvider Provider
+}
+
+// InitializeProvider returns a CAPI provider implementation for a specific
+// cloud platform.
+func InitializeProvider(platform Provider) infrastructure.Provider {
+	return InfraProvider{capiProvider: platform}
+}
+
+// CAPIInfraHelper provides an interface for calling functions at different
+// points of the CAPI infrastructure provisioning lifecycle.
+type Provider interface {
+	// PreProvision is called before provisioning using CAPI controllers has begun.
+	// and should be used to create dependencies needed for CAPI provisioning,
+	// such as IAM roles or policies.
+	PreProvision(in PreProvisionInput) error
+
+	// Ignition generates the ignition secrets for bootstrap and control-plane machines
+	// and handles any preconditions for ignition.
+	Ignition(in IgnitionInput) ([]client.Object, error)
```

Comment:
Based on discussions with OpenStack team (and reviewing #7856), I am floating the idea of adding a function to allow cloud-specific creation of the ignition secrets.  You can see the implementation in the defaultCAPIProvider.

Ultimately, I do think this logic should live in the CAPI provider but we should add this function here and take on the tech debt. We would want this functionality to live in the CAPI provider, so that the behavior would be the same if we were using a management cluster to provision the infrastructure. 

@vincepri wdyt?

Reply:
Sounds good to me, we can add documentation that this will have to move somewhere else